src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,567 @@
-    rw C_0 at h1, rw pow_one at h2,
-    rw [finset.prod_insert hys, nat_degree_mul (mul_ne_zero h1 h2), ih hx.2,
-        finset.card_insert_of_not_mem hys, nat_degree_mul h1 h2,
-        nat_degree_C, zero_add, nat_degree, degree_X_sub_C, add_comm], refl,
-    rw [ne, finset.prod_eq_zero_iff], rintro ⟨z, hzs, hz⟩,
-    rw mul_eq_zero at hz, cases hz with hz hz,
-    { rw [← C_0, C_inj, inv_eq_zero, sub_eq_zero] at hz, exact hx.2 (hz.symm ▸ hzs) },
-    { rw ← pow_one (X : F[X]) at hz, exact X_pow_sub_C_ne_zero zero_lt_one _ hz } }
+  simp_rw [basis_divisor, mul_eq_zero, X_sub_C_ne_zero, or_false,
+            C_eq_zero, inv_eq_zero, sub_eq_zero] at hxy,
+  exact hxy
 end
 
-variables (f : F → F)
+@[simp] lemma basis_divisor_eq_zero_iff : basis_divisor x y = 0 ↔ x = y :=
+⟨basis_divisor_injective, λ H, H ▸ basis_divisor_self_zero⟩
+
+lemma basis_divisor_ne_zero_iff : basis_divisor x y ≠ 0 ↔ x ≠ y :=
+by rw [ne.def, basis_divisor_eq_zero_iff]
+
+lemma degree_basis_divisor_ne (hxy : x ≠ y) : (basis_divisor x y).degree = 1 :=
+begin
+  rw [basis_divisor, degree_mul, degree_X_sub_C, degree_C, zero_add],
+  exact inv_ne_zero (sub_ne_zero_of_ne hxy)
+end
 
-/-- Lagrange interpolation: given a finset `s` and a function `f : F → F`,
-`interpolate s f` is the unique polynomial of degree `< s.card`
-that takes value `f x` on all `x` in `s`. -/
-def interpolate : F[X] :=
-∑ x in s, C (f x) * basis s x
+@[simp] lemma degree_basis_divisor_eq : (basis_divisor x x).degree = ⊥ :=
+by rw [basis_divisor_self_zero, degree_zero]
 
-@[simp] theorem interpolate_empty (f) : interpolate (∅ : finset F) f = 0 :=
-rfl
+lemma nat_degree_basis_divisor_eq : (basis_divisor x x).nat_degree = 0 :=
+by rw [basis_divisor_self_zero, nat_degree_zero]
 
-@[simp] theorem interpolate_singleton (f) (x : F) : interpolate {x} f = C (f x) :=
-by rw [interpolate, finset.sum_singleton, basis_singleton_self, mul_one]
+lemma nat_degree_basis_divisor_ne (hxy : x ≠ y) : (basis_divisor x y).nat_degree = 1 :=
+nat_degree_eq_of_degree_eq_some (degree_basis_divisor_ne hxy)
 
-@[simp] theorem eval_interpolate (x) (H : x ∈ s) : eval x (interpolate s f) = f x :=
+@[simp] lemma eval_basis_divisor_right : eval y (basis_divisor x y) = 0 :=
+by simp only [basis_divisor, eval_mul, eval_C, eval_sub, eval_X, sub_self, mul_zero]
+
+lemma eval_basis_divisor_left_ne (hxy : x ≠ y) : eval x (basis_divisor x y) = 1 :=
+begin
+  simp only [basis_divisor, eval_mul, eval_C, eval_sub, eval_X],
+  exact inv_mul_cancel (sub_ne_zero_of_ne hxy)
+end
+
+end basis_divisor
+
+section basis
+open finset
+variables {ι : Type*} [decidable_eq ι] {s : finset ι} {v : ι → F} {i j : ι}
+
+/-- Lagrange basis polynomials indexed by `s : finset ι`,
+defined at nodes `v i` for a map `v : ι → F` For `i, j ∈ s`, `basis s v i` evaluates to 1 at
+`v i` and 0 at `v j` for `i ≠ j`. -/
+def basis (s : finset ι) (v : ι → F) (i : ι) : F[X] := ∏ j in s.erase i, basis_divisor (v i) (v j)
```

Comment:
This is probably going to conflict with `basis` from linear algebra, so unless it gives too much noise let's make it `protected`.
```suggestion
protected def basis (s : finset ι) (v : ι → F) (i : ι) : F[X] := ∏ j in s.erase i, basis_divisor (v i) (v j)
```